### PR TITLE
[FLINK-15487][table] Allow registering FLIP-65 functions in TableEnvironment

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This updates all catalog related interfaces to FLIP-65. It also continues FLIP-64 by exposing new interfaces in table environment for registering functions of temporary/system/catalog kind. Furthermore, it adds early class validation to the function catalog.

The first functions will be fully functional once the code generator has been updated.

## Brief change log

- Update of `TableEnvironment`
- Update of `FunctionCatalog`
- Update of `FunctionCatalogOperatorTable`

## Verifying this change

- New `FunctionCatalogTest` cases
- New `UserDefinedFunctionHelperTest` case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+			UserDefinedFunctionHelper.prepareInstance(new TableConfig(), testSpec.functionInstance);
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	// Test utilities
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Test specification shared with the Scala tests.
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java]: nit: Is this comment valid?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java [hunk 2]
```diff
@@ -81,8 +85,283 @@ public void setPlannerTypeInferenceUtil(PlannerTypeInferenceUtil plannerTypeInfe
...
+		}
+
+		return definition != null;
+	}
+
+	/**
+	 * Registers a temporary catalog function.
+	 */
+	public void registerTemporaryCatalogFunction(
+			UnresolvedIdentifier unresolvedIdentifier,
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java]: We're not being too consistent with the type of identifier that the `FunctionCatalog/CatalogManager` accepts (I'm guilty of that myself).

Some methods accept `UnresolvedIdentifier` e.g. `FunctionCatalog#registerTemporaryCatalogFunction`, `CatalogManager#dropTemporaryView`. Some resolved `ObjectIdentifier` e.g. `CatalogManager#createTemporaryTable`, `CatalogManager#createTable`.

I am not sure which one should we prefer. If we go with the `UnresolvedIdentifier` the benefit is that we always qualify in the `Catalog*`. The downside is that we would no longer qualify in the `*Operations`, (e.g. CreateTableOperation etc.), whereas we said that all Operations should be fully resolved...

We can create a follow up for that.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 3]
```diff
@@ -273,6 +274,72 @@ public void registerFunction(String name, ScalarFunction function) {
 			function);
 	}
 
+	@Override
+	public void createTemporarySystemFunction(String name, Class<? extends UserDefinedFunction> functionClass) {
+		final UserDefinedFunction functionInstance = UserDefinedFunctionHelper.instantiateFunction(functionClass);
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: Yes, that makes sense. We can do this while integrating the `FunctionCatalog` into `CatalogManager`.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java [hunk 4]
```diff
@@ -97,7 +97,13 @@
 
 		@Override
 		public List<ResolvedExpression> visit(UnresolvedCallExpression unresolvedCall) {
-			final FunctionDefinition definition = prepareUserDefinedFunction(unresolvedCall.getFunctionDefinition());
+			final FunctionDefinition definition;
+			// clean functions that were not registered in a catalog
+			if (!unresolvedCall.getFunctionIdentifier().isPresent()) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Yes, this is a shortcoming. I will create a follow up issue for this.

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: https://issues.apache.org/jira/browse/FLINK-15804

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java [hunk 5]
```diff
@@ -238,36 +244,40 @@ private ResolvedExpression runLegacyTypeInference(
...
 				final TableAggregateFunctionDefinition taf = (TableAggregateFunctionDefinition) definition;
-				UserDefinedFunctionHelper.prepareFunction(resolutionContext.configuration(), taf.getTableAggregateFunction());
+				UserDefinedFunctionHelper.prepareInstance(resolutionContext.configuration(), taf.getTableAggregateFunction());
 				return new TableAggregateFunctionDefinition(
 					taf.getName(),
 					taf.getTableAggregateFunction(),
 					taf.getResultTypeInfo(),
 					taf.getAccumulatorTypeInfo());
+			} else if (definition instanceof UserDefinedFunction) {
+				UserDefinedFunctionHelper.prepareInstance(
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/resolver/rules/ResolveCallByArgumentsRule.java]: Not yet :(

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java [hunk 6]
```diff
@@ -147,39 +157,156 @@
...
+	/**
+	 * Validates the implementation methods such as {@code eval()} or {@code accumulate()} depending
+	 * on the {@link UserDefinedFunction} subclass.
+	 *
+	 * <p>This method must be kept in sync with the code generation requirements and the individual
+	 * docs of each function.
+	 */
+	private static void validateImplementationMethods(Class<? extends UserDefinedFunction> functionClass) {
+		if (ScalarFunction.class.isAssignableFrom(functionClass)) {
+			validateImplementationMethod(functionClass, false, false, "eval");
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java]: I think this helper class is a good location. Will update the PR.

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java]: nope copy paste error

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java (hunk 2 above)
Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java]: I also observed these inconsistencies. I think the catalog manager can resolve identifiers internally and thus can just expose `UnresolvedIdentifier`. That's why all my newly added methods take the unresolved one.

We said that operation should be fully resolved but that was before the DDL. So things might have changed a bit. Maybe we should distinguish between resolved and unresolved operations. Let's discuss this in a follow up issue.

**Comment by twalthr** [comment]
@flinkbot run travis

**Comment by twalthr** [comment]
@dawidwys I addressed your feedback and Travis gives green light. If you have no objections, I will merge this. Can you take care of opening follow up issues? Thanks.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java (hunk 2 above)
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java]: Created https://issues.apache.org/jira/browse/FLINK-15859

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java]: Created https://issues.apache.org/jira/browse/FLINK-15860

**Pull request closed without merging**